### ISSUE TITLE
Graph Specification file may model types of data not present in some Instruction Set files?

### ISSUE THREAD

**Issue by greenguy33** [PennTURBO/Drivetrain#21]
Background: The `Graph Specification` file represents the final state of our expanded graph. In theory there should be only one `Graph Specification`. Numerous `Instruction Set` files can be created but each must fulfill every `Connection Recipe` referenced in the `Graph Specification`. There is validation in place to enforce that this happens, and an `Instruction Set` is declared invalid otherwise, with the following error message:
```
Error in graph model: connection *someConnection* in graph specification is not the output of a queued process in the data model
```
@turbomam is currently working on what will become our single `Graph Specification` file, using the data fields from his OMOP database as guidelines. When this is finished he will pass it off to me and I will add the components required for PDS which do not exist in OMOP, namely `Biobank Encounters` and `Loss Of Function Genomic Data`. 

However this may pose a problem. Once I have added the extra fields necessary for an instantiation from PDS, Mark would then be unable to run an instantiation from his OMOP instruction set, due to the aforementioned validation, unless he also modeled the extra PDS fields in his instruction set. Asking a user to model data fields which don't exist in their specific data sets does not seem like a good practice going forwards, so it seems like a better solution will become necessary.

I'll spend some time this week trying to think of a solution.

**Comment by hjwilli** [comment]
Hi Hayden, all the triple recipes in the graph specification do *not* necessarily need to be the output of processes in a given instruction set.  If there are triples that should or could exist when instantiating from PDS/carnival but will definitely not when instantiating from raw OMOP tables, then both the instruction sets you listed above would be valid and neither should throw an error.   The recipes that are always required should be represented in the `Graph Specification` in a more fine-grained way.  It's very good that you have validation to check and log recipes are not outputs of an instruction set, but it should probably be a warning not an error.  

When we were initially spec-ing Acorn out we had a discussion about how to represent requirement/dependency validation in the triple recipes listed in the `Graph Specification`.  We came up with a few attributes that can be applied to a triple recipe to represent this (take below examples with a grain of salt, they are just meant to be illustrative)... 
- always required, meaning a triple represented by recipe must always exist (an example might be the recipe representing `patient -[is a]-> homosapien`... a valid drivetrain graph must have human patients)
- triple generated by this recipe is required if the subject exists (an example might be  `cancerous tumor-[is associated with]-> a cancer disposition`... if there is a cancer tumor instance it must be associated with a cancer disposition, but perhaps it is valid to have a cancer disposition by other means, like implied by ICD codes)
- triple generated by this recipe is required if the object exists (an example might be `healthcare encounter-[mentions]->prescription`... if there is a prescription it must be associated with a healthcare encounter, but a healthcare encounter can exist without  prescriptions)

If a recipe has been marked as always required and it is not the output of a process in an `instruction set`, that should definitely cause an error when validating the instruction set.

**Comment by greenguy33** [comment]
Hey Heather, thanks for the in depth response! It's noted that you think the validation in this case should throw a warning, not an error. I also would like to take a moment to clarify what you mean when you are talking about requirements for connection recipes.

Currently there are 4 different flavors of requirement for a connection recipe: `bothRequired`, `notRequired`, `subjectRequired`, and `objectRequired`. These are currently used to validate the input data.  However, I think this is actually a separate issue from what kinds of data are included.

First, I should say that it is not the case that a Drivetrain instantiation must have any specific kind of data. The program can expand just biobank encounters, or healthcare encounters, or patients, and none of them are dependent on any other. It is not necessary (currently) to have patients in the instantiation. So there is actually no triple that is explicitly required and must exist in all cases. 

The `bothRequired` flag for a connection recipe means that has properties of both `subjectRequired` and `objectRequired`, that is to say that if the subject of the recipe is present the object must be present and if the object is present then the subject must be present. However, if we assume this connection is between, for example, a biobank encounter CRID and a biobank encounter, and our incoming data set does not have any biobank encounter data, this is still a valid state for our graph.

As such I do not think that using Acorn's `required` property is suitable for validating a `Graph Specification` against an `Instruction Set`.

**Comment by hjwilli** [comment]
Hey @Hayden, to clarify... does `subjectRequired` mean that if an instance of the subject exists in the final graph exists then a triple represented by this recipe must also exist in the final graph?  Or does it mean something else?

**Comment by greenguy33** [comment]
@hjwilli actually, it means that if an instance of the _object_ exists in the final graph, then a triple represented by the recipe must also exist. That is to say, if an instance of this object exists, then an instance of the subject is required, hence `subjectRequired`

The problem is that if an instance of the object does not exist, then we cannot make any declarations about whether there will be an instance of the subject or not

**Comment by hjwilli** [comment]
Gotcha, thanks.  If I'm understanding you correctly, then that naming convention seems a little confusing... if an instance of the object exists it doesn't *just* mean that an instance of the subject exists, it means the whole triple exists!  (an instance of the subject might exist by other means).  I was thinking more along the lines of "requirement for this triple recipe to run".  Regardless, I think I understand your logical definition now.

You can still use the dependency structure implied by `subjectRequired` `objectRequired` or `bothRequired` to do a first pass of what recipes need to be in a valid instruction set.

It's true that if an instance of the object does not exist, then we cannot make any declarations about whether there will be an instance of the subject or not.  However, if an instance of an object *might* exist (because there is a recipe with that class as either subject or object in the instruction set), then any recipes that trigger off of instances of that class (the subject of an `objectRequired` recipe or the object of a `subjectRequired` recipe, subject/object of `bothRequired` recipe) must also be in the instruction set.  If they aren't, it means the instruction set is capable of producing graphs that violate the `GraphSpec`.

On the flip side- if your graph spec has a bunch of recipes that have, say, 'specimen' as a required thing (subject or object depending on the requirement type), and you have an instruction set has no recipes that involve the 'specimen' class but otherwise passes the dependency validation above, then it would be fine.

I think the information we need to validate an instruction set is already contained in the GraphSpec, we just need to figure out how to use it.

**Comment by hjwilli** [comment]
tldr; 
if a recipe in the graphSpec is not in an instruction set, throw a warning.

get all classes of the subjects and objects of the recipes in the instruction set.  If there are any recipes in the graphSpec that must exist if there are instances of those classes, but those recipes are not in the instruction set, throw an error.

**Comment by greenguy33** [comment]
I'm still having trouble seeing how this solves the problem of data which is modeled in the graph spec but not relevant to a user making an instruction set. However perhaps I am missing something. I will draw up something soon to illustrate an example where I think this could break down, and you can tell me if I am misunderstanding you.

**Comment by hjwilli** [comment]
followed up in person with @greenguy33 @augustearth during the dev meeting.

after a clarifying discussion, we agreed that the path outlined in the tl;dr should solve the issue @greenguy33 outlined in the opening post.

The requirements nomenclature for recipes will also be updated- recipes will have an attribute along the lines of `mustExistIf` that can have values like `subjectExists`, `objectExists`, `eitherSubjectOrObjectExists`
